codice/ddf
# DDF-5430 Add ruler capabilites to right-click context menu on map

#### What does this PR do?

Adds the full implementation for the Intrigue ruler feature brought up during the UC Hackathon. A new menu action is added to the right-click context menu that contains options for using the ruler on both the 2D and 3D Map. Coordinates can be selected as start/end points for the ruler and the distance between them is calculated. Map markers are placed at the selected coordinates and a line is drawn between them to show the distance that is being measured. The map markers and line are able to be cleared through a menu option or when new coordinates are selected after a distance has been calculated for the previous set.

#### Who is reviewing it? 

@rfding 
@aaronilovici
@josephthweatt 
@brjeter 
@pvargas 

#### Select relevant component teams: 

@codice/ui 

#### Ask 2 committers to review/merge the PR and tag them here.

@andrewkfiedler
@djblue 

#### How should this be tested?

1. Open the Intrigue UI and go to the 3D Map view.
2. Right-click anywhere on the map to bring up the context menu.
3. Click on the `Measure Distance` dropdown menu.
4. Click on `Select start point` and verify that a map marker labeled `A` is placed on the selected coordinate.
5. Repeat steps 2-3 for a different point on the map.
6. Click on `Select end point` and verify that both a map marker labeled `B` is placed on the selected coordinate and a line is drawn between the two markers.
7. Verify that an announcement is displayed with a distance measurement.
8. Repeat steps 2-7 for the 2D Map view.

#### Any background context you want to provide?
@pvargas worked on the original version for the UC Hackathon. This pull request was worked on as a collaboration with @rfding.

#### What are the relevant tickets?
Fixes: #5430 #5473 

#### Screenshots

Menu actions:
![image](https://user-images.githubusercontent.com/10703203/67963434-fdba6300-fbd4-11e9-97b9-22bcab091d67.png)

Distance measurement display:
<img width="1460" alt="Screen Shot 2019-10-31 at 11 50 05 AM" src="https://user-images.githubusercontent.com/10703203/67963457-0448da80-fbd5-11e9-82fd-e1557d6cd2a4.png">

#### Checklist:

#### Notes on Review Process
Please see [Notes on Review Process](https://codice.atlassian.net/wiki/spaces/DDF/pages/71946981/Pull+Request+Guidelines) for further guidance on requirements for merging and abbreviated reviews. 

#### Review Comment Legend:
- ✏️ (Pencil) This comment is a nitpick or style suggestion, no action required for approval. This comment should provide a suggestion either as an in line code snippet or a gist. 
- ❓ (Question Mark) This comment is to gain a clearer understanding of design or code choices, clarification is required but action may not be necessary for approval.
- ❗ (Exclamation Mark) This comment is critical and requires clarification or action before approval.
